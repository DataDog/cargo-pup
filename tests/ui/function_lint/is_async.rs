// This product includes software developed at Datadog (https://www.datadoghq.com/) Copyright 2024 Datadog, Inc.

//@compile-flags: --crate-name test_is_async
//@compile-flags: --crate-type lib

// This test verifies that the FunctionLint's IsAsync matcher works correctly

// Async function that should trigger the lint
async fn async_function() { //~ ERROR: Function 'async_function' is forbidden by lint rule
}

// Async function with return type
async fn async_with_return() -> String { //~ ERROR: Function 'async_with_return' is forbidden by lint rule
    "hello".to_string()
}

// Regular function that should NOT trigger the lint
fn sync_function() {
}

// Another regular function with return type
fn sync_with_return() -> String {
    "hello".to_string()
}

// Async method in impl block
struct TestStruct;

impl TestStruct {
    async fn async_method(&self) { //~ ERROR: Function 'async_method' is forbidden by lint rule
    }
    
    fn sync_method(&self) {
    }
}

// Async functions in traits
trait AsyncTrait {
    async fn trait_async_method(&self);
    
    fn trait_sync_method(&self);
}

impl AsyncTrait for TestStruct {
    async fn trait_async_method(&self) { //~ ERROR: Function 'trait_async_method' is forbidden by lint rule
    }
    
    fn trait_sync_method(&self) {
    }
}