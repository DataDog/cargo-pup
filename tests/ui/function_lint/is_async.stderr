error: Function 'async_function' is forbidden by lint rule
  --> tests/ui/function_lint/is_async.rs:9:1
   |
LL | async fn async_function() {
   | ^^^^^^^^^^^^^^^^^^^
   |
   = help: Remove this function to satisfy the architectural rule
   = note: Applied by cargo-pup rule 'async_function_test'.
   = note: `#[deny(function_lint)]` on by default

error: Function 'async_with_return' is forbidden by lint rule
  --> tests/ui/function_lint/is_async.rs:13:1
   |
LL | async fn async_with_return() -> String {
   | ^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Remove this function to satisfy the architectural rule
   = note: Applied by cargo-pup rule 'async_function_test'.

error: Function 'async_method' is forbidden by lint rule
  --> tests/ui/function_lint/is_async.rs:30:5
   |
LL |     async fn async_method(&self) {
   |     ^^^^^^^^^^^^^^^^^
   |
   = help: Remove this function to satisfy the architectural rule
   = note: Applied by cargo-pup rule 'async_function_test'.

error: Function 'trait_async_method' is forbidden by lint rule
  --> tests/ui/function_lint/is_async.rs:45:5
   |
LL |     async fn trait_async_method(&self) {
   |     ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Remove this function to satisfy the architectural rule
   = note: Applied by cargo-pup rule 'async_function_test'.

error: aborting due to 4 previous errors

